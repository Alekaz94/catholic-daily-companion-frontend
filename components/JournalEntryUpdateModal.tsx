import { useEffect, useState } from "react";
import { JournalEntry, UpdateJournalEntry } from "../models/JournalEntry"
import { Modal, TextInput, View, Text, Button, StyleSheet } from "react-native";

interface Props {
    visible: boolean,
    entry: JournalEntry | null;
    onClose: () => void;
    onUpdate: (id: string, updatedEntry: UpdateJournalEntry) => void;
}

const JournalEntryUpdateModal: React.FC<Props> = ({visible, entry, onClose, onUpdate}) => {
    const [title, setTitle] = useState("");
    const [content, setContent] = useState("");

    useEffect(() => {
        if(entry) {
            setTitle(entry.title)
            setContent(entry.content)
        }
    }, [entry])

    const onHandleSubmit = () => {
        if(!entry) {
            return;
        }

        const trimmedTitle = title.trim();
        const trimmedContent = content.trim();

        if(!trimmedTitle && !trimmedContent) {
            alert("Both title and content are empty. No changes saved.");
            return;
        }

        const updatedFields: UpdateJournalEntry = {};

        if(trimmedTitle && trimmedTitle !== entry.title) {
            updatedFields.title = trimmedTitle;
        }

        if(trimmedContent && trimmedContent !== entry.content) {
            updatedFields.content = trimmedContent;
        }

        if (Object.keys(updatedFields).length === 0) {
            alert("No changes detected.");
            return;
        }

        onUpdate(entry.id, updatedFields);
        onClose();
    }

    if(!entry) {
        return null;
    }

    return (
        <Modal visible={visible} animationType="slide">
            <View style={styles.container}>
                <Text style={styles.title}>Edit Entry</Text>
                    <TextInput 
                        placeholder="Enter title..."
                        style={styles.input}
                        value={title}
                        onChangeText={(value) => setTitle(value)}
                    />
                    <TextInput 
                        placeholder="Write your journal entry..."
                        style={styles.multiline}
                        value={content}
                        onChangeText={(value) => setContent(value)}
                    /> 
                    <Button title="Save changes" onPress={onHandleSubmit} /> 
                    <Button title="Cancel" color="gray" onPress={onClose} />
            </View>
        </Modal>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 20
    },
    title: {
        fontSize: 20,
        fontWeight: "bold",
        marginBottom: 16
    },
    input: {
        borderWidth: 1,
        padding: 10,
        marginBottom: 12,
        borderRadius: 6
    },
    multiline: {
        borderWidth: 1,
        padding: 10,
        marginBottom: 12,
        borderRadius: 6,
        height: 200,
        textAlignVertical: "top"
    }
})

export default JournalEntryUpdateModal;