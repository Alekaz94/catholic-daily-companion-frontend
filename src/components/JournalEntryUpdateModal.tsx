import { useEffect, useState } from "react";
import { JournalEntry, UpdateJournalEntry } from "../models/JournalEntry"
import { Modal, TextInput, View, Text, TouchableOpacity, ActivityIndicator, TouchableWithoutFeedback, Keyboard } from "react-native";
import { Layout } from "../styles/Layout";
import { Typography } from "../styles/Typography";
import { AppTheme } from "../styles/colors";
import React from "react";

interface Props {
    visible: boolean,
    entry: JournalEntry | null;
    onClose: () => void;
    onUpdate: (id: string, updatedEntry: UpdateJournalEntry) => Promise<void>;
}

const JournalEntryUpdateModal: React.FC<Props> = ({visible, entry, onClose, onUpdate}) => {
    const [title, setTitle] = useState("");
    const [content, setContent] = useState("");
    const [isLoading, setIsLoading] = useState(false);

    useEffect(() => {
        if(entry) {
            setTitle(entry.title)
            setContent(entry.content)
        }
    }, [entry])

    const onHandleSubmit = async () => {
        if(!entry) {
            return;
        }

        try {
            setIsLoading(true);
            const trimmedTitle = title.trim();
            const trimmedContent = content.trim();
        
            if(!trimmedTitle && !trimmedContent) {
                alert("Both title and content are empty. No changes saved.");
                return;
            }

            const updatedFields: UpdateJournalEntry = {};
        
            if(trimmedTitle && trimmedTitle !== entry.title) {
                updatedFields.title = trimmedTitle;
            }
        
            if(trimmedContent && trimmedContent !== entry.content) {
                updatedFields.content = trimmedContent;
            }
        
            if (Object.keys(updatedFields).length === 0) {
                alert("No changes detected.");
                return;
            }

            await onUpdate(entry.id, updatedFields);
            alert("Update successfull!");
            onClose();
        } catch (error) {
            alert("Update failed. Please try again.");
            console.error(error);
        } finally {
            setIsLoading(false);
        }
    }

    if(!entry) {
        return null;
    }

    return (
        <Modal visible={visible} animationType="slide">
            <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
                <View style={[Layout.container, {backgroundColor: AppTheme.journal.background}]}>
                    <Text style={[Typography.title, {color: AppTheme.journal.text}]}>Edit Entry</Text>
                    <TextInput 
                        editable={!isLoading}
                        placeholder="Enter title..."
                        style={Layout.input}
                        value={title}
                        onChangeText={(value) => setTitle(value)}
                    />
                    <TextInput 
                        editable={!isLoading}
                        placeholder="Write your journal entry..."
                        style={[Layout.input, {height: 200, textAlignVertical: "top"}]}
                        value={content}
                        onChangeText={(value) => setContent(value)}
                    /> 
                    <View style={{flexDirection: "row", justifyContent: "space-between"}}>
                        <TouchableOpacity style={[Layout.button, {width: "40%", alignSelf: "center", backgroundColor: "#B794F4", borderWidth: 1, opacity: isLoading ? 0.7 : 1}]} onPress={onHandleSubmit} >
                            {isLoading ? (
                                <ActivityIndicator color="black" />
                            ) : (
                                <Text style={[Layout.buttonText, {color: AppTheme.journal.text}]}>Save changes</Text>
                            )}
                        </TouchableOpacity>
                        
                        <TouchableOpacity disabled={isLoading} style={[Layout.button, {backgroundColor: "gray", width: "40%", alignSelf: "center", borderWidth: 1, opacity: isLoading ? 0.7 : 1}]} onPress={onClose} >
                            <Text style={Layout.buttonText}>Cancel</Text>
                        </TouchableOpacity>
                    </View>
                </View>
            </TouchableWithoutFeedback>
        </Modal>
    );
}

export default JournalEntryUpdateModal;