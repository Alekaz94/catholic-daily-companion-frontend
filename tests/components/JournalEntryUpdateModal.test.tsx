import { act, fireEvent, render } from "@testing-library/react-native"
import JournalEntryUpdateModal from "../../components/JournalEntryUpdateModal"

const mockJournalEntry = {
    id: "1sad23",
    date: "2025-08-25",
    updatedAt: "2025-08-30",
    title: "Test entry",
    content: "This is the content of the test entry"
}

const mockUpdatedJournalEntry = {
    title: "Updated Entry Title",
    content: "Updated content to test entry"
}

describe("JournalEntryUpdateModal", () => {
    it("does not render when entry is null", () => {
        const { toJSON } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={null}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />
        )
        expect(toJSON()).toBeNull();
    });

    it("does not render content if visible is false", () => {
        const { queryByText } = render(
            <JournalEntryUpdateModal 
                visible={false}
                entry={mockJournalEntry}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />        
        );
        expect(queryByText("Test entry")).toBeNull();
    });

    it("renders input fields with entry data", () => {
        const { getByDisplayValue } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />
        );
        expect(getByDisplayValue("Test entry")).toBeTruthy();
        expect(getByDisplayValue("This is the content of the test entry")).toBeTruthy();
    })

    it("calls onClose when Close button is pressed", async () => {
        const onCloseMock = jest.fn();
        const { getByText } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={onCloseMock}
                onUpdate={jest.fn()}
            />
        )
        
        await act(async () => {
            fireEvent.press(getByText("Cancel"));
        })
        expect(onCloseMock).toHaveBeenCalled();
    })

    it("calls onUpdate and onClose with trimmed, changed values", async () => {
        globalThis.alert = jest.fn();
        const onCloseMock = jest.fn();
        const onUpdateMock = jest.fn();
        const { getByText, getByPlaceholderText } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={onCloseMock}
                onUpdate={onUpdateMock}
            />
        )

        await act( async () => {
            fireEvent.changeText(getByPlaceholderText("Enter title..."), "Updated Entry Title");
            fireEvent.changeText(getByPlaceholderText("Write your journal entry..."), "Updated content to test entry");
        })
        expect(getByPlaceholderText("Enter title...").props.value).toBe("Updated Entry Title");
        expect(getByPlaceholderText("Write your journal entry...").props.value).toBe("Updated content to test entry");

        await act(async () => {
            fireEvent.press(getByText("Save changes"));
        })
        expect(globalThis.alert).toHaveBeenCalledWith("Update successful!");
        expect(onUpdateMock).toHaveBeenCalledWith(mockJournalEntry.id, mockUpdatedJournalEntry);
        expect(onCloseMock).toHaveBeenCalled();
    })

    it("shows alert and does not call onUpdate if no changes are made", async () => {
        globalThis.alert = jest.fn();

        const { getByText } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />
        )

        await act(async () => {
            fireEvent.press(getByText("Save changes"));
        })
        expect(globalThis.alert).toHaveBeenCalledWith("No changes detected.")
    })

    it("shows alert if both title and content fields are empty", async () => {
        globalThis.alert = jest.fn();

        const { getByText, getByPlaceholderText } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />
        )
        const titleInput = getByPlaceholderText("Enter title...");
        const contentInput = getByPlaceholderText("Write your journal entry...");

        await act(async () => {
            fireEvent.changeText(titleInput, "");
        })

        await act(async () => {
            fireEvent.changeText(contentInput, "");
        })
        await act(async () => {
            fireEvent.press(getByText("Save changes"));
        })
        expect(globalThis.alert).toHaveBeenCalledWith("Both title and content are empty. No changes saved.")
    })

    it("matches the snapshot", () => {
        const { toJSON } = render(
            <JournalEntryUpdateModal 
                visible={true}
                entry={mockJournalEntry}
                onClose={jest.fn()}
                onUpdate={jest.fn()}
            />        
        );
        expect(toJSON()).toMatchSnapshot();
    });
})