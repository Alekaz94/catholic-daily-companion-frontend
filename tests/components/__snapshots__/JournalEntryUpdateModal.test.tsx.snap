// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalEntryUpdateModal matches the snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  visible={true}
>
  <View
    style={
      [
        {
          "flex": 1,
          "padding": 20,
        },
        {
          "backgroundColor": "#EDE9FE",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#222",
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 16,
          },
          {
            "color": "#1A1A1A",
          },
        ]
      }
    >
      Edit Entry
    </Text>
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter title..."
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#dddddd",
          "borderRadius": 6,
          "borderWidth": 1,
          "marginBottom": 12,
          "padding": 10,
        }
      }
      value="Test entry"
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Write your journal entry..."
      style={
        [
          {
            "backgroundColor": "#fff",
            "borderColor": "#dddddd",
            "borderRadius": 6,
            "borderWidth": 1,
            "marginBottom": 12,
            "padding": 10,
          },
          {
            "height": 200,
            "textAlignVertical": "top",
          },
        ]
      }
      value="This is the content of the test entry"
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#B794F4",
            "borderRadius": 8,
            "marginTop": 10,
            "opacity": 1,
            "paddingVertical": 12,
            "width": "40%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "white",
                "fontSize": 16,
                "fontWeight": "bold",
              },
              {
                "color": "#1A1A1A",
              },
            ]
          }
        >
          Save changes
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "gray",
            "borderRadius": 8,
            "marginTop": 10,
            "opacity": 1,
            "paddingVertical": 12,
            "width": "40%",
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
